### PR TITLE
Exponential backoff & error handling for metadata validation

### PR DESCRIPTION
**Pull request by bistline**
Adding exponential backoff with "[full jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)" to all external calls to OLS when validating convention metadata.  Caching of retrieved ontologies has also been added to reduce the number of external API calls and speed up validation.  If all retries fail to connect (and validation is therefore impossible at the current time) then an error is reported back to ingest with a special error message denoting that an 'External service outage' has occurred.  This message will then be included in an email back to the user from the portal.

This PR satisfies [SCP-2076](https://broadworkbench.atlassian.net/browse/SCP-2076).

### CODE REVIEW

File: ingest/validation/validate_metadata.py [hunk 1]
```diff
@@ -493,27 +498,45 @@ def exit_if_errors(metadata):
         exit(1)
     return errors
 
-
+def backoff_hdlr(details):
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: Yeah that was lifted from their examples in the backoff repo.  Changing now.

File: ingest/validation/validate_metadata.py [hunk 2]
```diff
@@ -522,15 +545,30 @@ def retrieve_ontology_term(convention_url, ontology_id):
     # when ontologyID value is empty string -> TypeError
     except (ValueError, TypeError):
         return None
-    metadata_url = OLS_BASE_URL + ontology_shortname
-    metadata_ontology = retrieve_ontology(metadata_url)
+    # check if we have already retrieved this ontology reference
+    if ontology_shortname not in ontologies.keys():
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: Ah - I didn't know you could just pass the dictionary and do a key presence check.

File: ingest/validation/validate_metadata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: [9aac17c](https://github.com/broadinstitute/scp-ingest-pipeline/pull/76/commits/9aac17c48136846f1a44274fd43f37c44a0e57b6)

File: ingest/validation/validate_metadata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: [9aac17c](https://github.com/broadinstitute/scp-ingest-pipeline/pull/76/commits/9aac17c48136846f1a44274fd43f37c44a0e57b6)

File: ingest/validation/validate_metadata.py [hunk 3]
```diff
@@ -553,6 +591,11 @@ def encode_term_iri(term_id, base_uri):
     encoded_iri = encoder.quote_plus(encoder.quote_plus(query_uri))
     return encoded_iri
 
+def extract_terminal_pathname(url):
+    """Extract the last path segment from a URL
+    """
+    return list(filter(None, url.split("/"))).pop()
```
Comment (devonbush) [inline ingest/validation/validate_metadata.py]: the rails URI gem might have this built in?

File: ingest/validation/validate_metadata.py [hunk 4]
```diff
@@ -493,27 +498,45 @@ def exit_if_errors(metadata):
         exit(1)
     return errors
 
-
+def backoff_hdlr(details):
+    """Handler function to log backoff attempts when querying OLS"""
+    logger.debug("Backing off {wait:0.1f} seconds afters {tries} tries "
+           "calling function {target} with args {args} and kwargs "
+           "{kwargs}".format(**details))
+
```
[the remark this answers is not in the document]
Comment (bistline) [reply ingest/validation/validate_metadata.py]: It gets invoked from the `on_backoff=backoff_handler` parameter of the `@backoff` decorator.

File: ingest/validation/validate_metadata.py (hunk 3 above)
Comment (bistline) [reply ingest/validation/validate_metadata.py]: Except that this is Python...

**Pull request merged**